

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Installation guide &mdash; netZooC 0.2 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript">
          var DOCUMENTATION_OPTIONS = {
              URL_ROOT:'../',
              VERSION:'0.2',
              LANGUAGE:'None',
              COLLAPSE_INDEX:false,
              FILE_SUFFIX:'.html',
              HAS_SOURCE:  true,
              SOURCELINK_SUFFIX: '.txt'
          };
      </script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Functions" href="../functions/index.html" />
    <link rel="prev" title="Home" href="../index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> netZooC
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Home</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Installation guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#requirements">Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="#install">Install</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#panda">PANDA</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#data">DATA:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#file-formats">FILE FORMATS:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#outputted-network-files-including-stats-files">Outputted network files (including “stats” files):</a></li>
<li class="toctree-l4"><a class="reference internal" href="#additional-options">ADDITIONAL OPTIONS:</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#puma">PUMA</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#troubleshooting">Troubleshooting</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../functions/index.html">Functions</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutos/index.html">Tutorials</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">netZooC</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Installation guide</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/install/index.md.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="installation-guide">
<h1>Installation guide<a class="headerlink" href="#installation-guide" title="Permalink to this headline">¶</a></h1>
<p>To install netZooC on your computer, please check the following requirements:</p>
<div class="section" id="requirements">
<h2>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>g++ compiler</li>
</ul>
</div>
<div class="section" id="install">
<h2>Install<a class="headerlink" href="#install" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>git clone https://github.com/netZoo/netZooC.git</li>
<li>cd netZooC/zooAnimalFolder</li>
<li>make</li>
<li>Upon successful make, please run the program as detailed in the example section of the function description.</li>
</ul>
<div class="section" id="panda">
<h3>PANDA<a class="headerlink" href="#panda" title="Permalink to this headline">¶</a></h3>
<p>Written by Kimberly Glass (kglass&#64;jimmy.harvard.edu), available under GPL.  As academic code it is provided without warranty.  Please contact the author with any comments/questions/concerns.
Last updated July, 2014.</p>
<p>PANDA is written in c++ and can be compiled by:</p>
<p><code class="docutils literal notranslate"><span class="pre">g++</span> <span class="pre">PANDA_vc.c</span> <span class="pre">-o</span> <span class="pre">PANDA</span></code></p>
<p>To improve runtime speed, it is also highly encouraged to include an optimazation flag:</p>
<p><code class="docutils literal notranslate"><span class="pre">g++</span> <span class="pre">PANDA_v2.c</span> <span class="pre">-O3</span> <span class="pre">-o</span> <span class="pre">PANDA</span></code></p>
<p>Running the program without any parameters will return a usage function:</p>
<p>./PANDA</p>
<p>Usage ./PANDA</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">e</span> <span class="p">(</span><span class="n">required</span><span class="p">)</span> <span class="n">file</span> <span class="n">of</span> <span class="n">expression</span> <span class="n">values</span> <span class="p">(</span><span class="n">can</span> <span class="n">alternately</span> <span class="n">be</span> <span class="n">a</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">gene</span> <span class="n">names</span><span class="p">)</span>

<span class="o">-</span><span class="n">m</span> <span class="p">(</span><span class="n">required</span><span class="p">)</span> <span class="n">pair</span> <span class="n">file</span> <span class="n">of</span> <span class="n">motif</span> <span class="n">edges</span>

<span class="o">-</span><span class="n">p</span> <span class="p">(</span><span class="n">optional</span><span class="p">)</span> <span class="n">pair</span> <span class="n">file</span> <span class="n">of</span> <span class="n">PPI</span> <span class="n">edges</span>

<span class="o">-</span><span class="n">o</span> <span class="p">(</span><span class="n">optional</span><span class="p">)</span> <span class="n">tag</span> <span class="k">for</span> <span class="n">output</span> <span class="n">files</span>

<span class="o">-</span><span class="n">a</span> <span class="p">(</span><span class="n">optional</span><span class="p">)</span> <span class="n">value</span> <span class="n">to</span> <span class="n">be</span> <span class="n">used</span> <span class="k">for</span> <span class="n">update</span> <span class="n">variable</span><span class="p">,</span> <span class="n">alpha</span> <span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span> 
</pre></div>
</div>
<p>Additional options (see README):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">k</span> <span class="p">(</span><span class="n">optional</span><span class="p">)</span> <span class="n">kill</span> <span class="n">the</span> <span class="n">program</span> <span class="n">after</span> <span class="n">it</span> <span class="n">has</span> <span class="n">run</span> <span class="n">k</span> <span class="n">steps</span> <span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>

<span class="o">-</span><span class="n">n</span> <span class="p">(</span><span class="n">optional</span><span class="p">)</span> <span class="n">output</span> <span class="n">a</span> <span class="s2">&quot;stats&quot;</span> <span class="n">file</span> <span class="n">every</span> <span class="n">n</span> <span class="n">steps</span> <span class="p">(</span><span class="n">default</span><span class="p">,</span> <span class="n">no</span> <span class="n">stats</span> <span class="n">file</span><span class="p">)</span>

<span class="o">-</span><span class="n">w</span> <span class="p">(</span><span class="n">optional</span><span class="p">)</span> <span class="n">file</span> <span class="k">with</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">covariate</span> <span class="n">weights</span>

<span class="o">-</span><span class="n">l</span> <span class="p">(</span><span class="n">optional</span><span class="p">)</span> <span class="n">leave</span> <span class="n">out</span> <span class="n">the</span> <span class="n">lth</span> <span class="n">sample</span> <span class="n">when</span> <span class="n">building</span> <span class="n">the</span> <span class="n">network</span>

<span class="o">-</span><span class="n">j</span> <span class="p">(</span><span class="n">optional</span><span class="p">)</span> <span class="n">retain</span> <span class="n">only</span> <span class="n">j</span> <span class="n">samples</span> <span class="n">when</span> <span class="n">building</span> <span class="n">the</span> <span class="n">network</span>

<span class="o">-</span><span class="n">r</span> <span class="p">(</span><span class="n">optional</span><span class="p">)</span> <span class="n">randomization</span> <span class="n">options</span>

<span class="o">-</span><span class="n">s</span> <span class="p">(</span><span class="n">optional</span><span class="p">)</span> <span class="n">value</span> <span class="n">to</span> <span class="n">seed</span> <span class="n">the</span> <span class="n">random</span> <span class="n">number</span> <span class="n">generator</span> <span class="p">(</span><span class="n">defaults</span> <span class="n">to</span> <span class="n">system</span> <span class="n">time</span><span class="p">)</span>

<span class="o">-</span><span class="n">v</span> <span class="p">(</span><span class="n">optional</span><span class="p">)</span> <span class="n">verbose</span> <span class="n">output</span> <span class="n">options</span> 
</pre></div>
</div>
<p>There are two required parameters, the names for two input files: a tab-delimited text file with the gene expression values and a tab-delimited text file with motif information.  You may also choose to specify an string to generate output file names, the default value for this is “PANDA_prediction”.</p>
<p>examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">./</span><span class="n">PANDA</span> <span class="o">-</span><span class="n">e</span> <span class="n">ToyExpressionData</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">m</span> <span class="n">ToyMotifData</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">o</span> <span class="n">ToyOutput</span>

<span class="o">./</span><span class="n">PANDA</span> <span class="o">-</span><span class="n">e</span> <span class="n">ToyExpressionData</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">m</span> <span class="n">ToyMotifData</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">p</span> <span class="n">ToyPPIData</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">a</span> <span class="mf">0.25</span> <span class="o">-</span><span class="n">o</span> <span class="n">ToyOutput</span>
</pre></div>
</div>
<p>Running these examples will produce a network file: ToyOutput_FinalNetwork.pairs</p>
<div class="section" id="data">
<h4>DATA:<a class="headerlink" href="#data" title="Permalink to this headline">¶</a></h4>
<p>Within the “YeastData” directory, the file “YeastNetwork_allTFxGene” contains the final z-score edge weights for the networks analyzed in “Passing Messages between Biological Networks to Refine Predicted Interactions.”  This folder also contains the expression, motif and PPI data files used to generate the networks.  Please see the supplemental material of the publication for more information on this data.</p>
<p>The “ToyData” directory has some small ‘toy’ datafiles that can be used to quickly run/understand PANDA.</p>
</div>
<div class="section" id="file-formats">
<h4>FILE FORMATS:<a class="headerlink" href="#file-formats" title="Permalink to this headline">¶</a></h4>
<p>Expression data file: In the expression data file the first column must contain gene names, and each subsequent column should contain expression values for those genes across conditions/samples (see “ToyExpressionData.txt”).  Note that if fewer than three conditions are contained in the expression data file, PANDA will initialize the co-regulatory network to an identity matrix. This file can also contain header lines, so long as those lines are preceeded by a hashtag (#).</p>
<p>Motif data file: The motif data file contains three columns (see “ToyMotifData.txt”).  The first column should contain regulators, the second those regulator’s target genes, and the third, an initial weight to give the interactions (recommend 1).  Any potential regulator to gene interaction not specified in this file is given a default interaction weight of 0.  Any interaction that is multiply defined will be given a weight equal to the sum of the weights given to the instances of that interaction.  Note that PANDA will give an warning message if a gene in the motif file is not contained in the expression data file. This file cannot contain header lines.</p>
<p>Protein interaction file: The protein interaction file (optional) contains three columns (see “ToyPPIData.txt”).  The first two columns should contain a pair of regulators, and the third, an initial weight to give the interactions (recommend 1).  Note that the program will give an warning message if a regulator in this file is not contained in the motif data. This file cannot contain header lines.</p>
</div>
<div class="section" id="outputted-network-files-including-stats-files">
<h4>Outputted network files (including “stats” files):<a class="headerlink" href="#outputted-network-files-including-stats-files" title="Permalink to this headline">¶</a></h4>
<p>The network files contain four columns:</p>
<p><code class="docutils literal notranslate"><span class="pre">TF</span> <span class="pre">\t</span> <span class="pre">Gene</span> <span class="pre">\t</span> <span class="pre">Motif-prediction</span> <span class="pre">\t</span> <span class="pre">PANDA-prediction</span></code></p>
<p>The values in the fourth, “PANDA-prediction”, column, can loosely be interpreted as Z-scores.</p>
</div>
<div class="section" id="additional-options">
<h4>ADDITIONAL OPTIONS:<a class="headerlink" href="#additional-options" title="Permalink to this headline">¶</a></h4>
<p>PANDA allows the user to specify a number of options. Some of these have been implemented as we apply the PANDA algorithm to other systems.  Any comments/questions/suggestions are welcomes (email: kglass&#64;jimmy.harvard.edu).</p>
<p>“-a” allows a user to specify a different value for the update parameter, alpha.  Larger values of alpha causes quicker convergence, and often greater “mixing” of motif and non-motif edges.  Recommend values between 0.05 and 0.25. (Author note: The default of 0.1 was chosen since as ChIP-seq validation of human networks appears to show this gives the most accurate results. However for quicker analysis I strongly suggest starting with 0.25 as there is only a small loss in accuracy but an immense speed-up in time).</p>
<p>example: <code class="docutils literal notranslate"><span class="pre">./PANDA</span> <span class="pre">-e</span> <span class="pre">ToyExpressionData.txt</span> <span class="pre">-m</span> <span class="pre">ToyMotifData.txt</span> <span class="pre">-a</span> <span class="pre">0.25</span> <span class="pre">-o</span> <span class="pre">ToyOutput</span></code></p>
<p>“-k” sets the maximum number of iterations PANDA can run before exiting.</p>
<p>example: <code class="docutils literal notranslate"><span class="pre">./PANDA</span> <span class="pre">-e</span> <span class="pre">ToyExpressionData.txt</span> <span class="pre">-m</span> <span class="pre">ToyMotifData.txt</span> <span class="pre">-k</span> <span class="pre">2</span> <span class="pre">-o</span> <span class="pre">ToyOutput</span></code></p>
<p>“-n” tells PANDA to print the network every N steps, instead of just the final predicted network.</p>
<p>example: <code class="docutils literal notranslate"><span class="pre">./PANDA</span> <span class="pre">-e</span> <span class="pre">ToyExpressionData.txt</span> <span class="pre">-m</span> <span class="pre">ToyMotifData.txt</span> <span class="pre">-n</span> <span class="pre">5</span> <span class="pre">-o</span> <span class="pre">ToyOutput</span></code></p>
<p>“-w” triggers PANDA to calculate a weighted Pearson (in lieu of the standard Pearson correlation), using weights specified in a “covariate weight file.”  The file should contain a single column of numbers.  Each row in the file corresponds directly to a column in the expression data file.</p>
<p>example usage: <code class="docutils literal notranslate"><span class="pre">./PANDA</span> <span class="pre">-e</span> <span class="pre">ToyExpressionData.txt</span> <span class="pre">-m</span> <span class="pre">ToyMotifData.txt</span> <span class="pre">-w</span> <span class="pre">ToyWeights.txt</span> <span class="pre">-o</span> <span class="pre">ToyOutput</span></code></p>
<p>“-l” can be used to leave out a specified expression sample when building the network. 1 leaves out the first sample, 2 leaves out the second sample, etc.</p>
<p>example usage: <code class="docutils literal notranslate"><span class="pre">./PANDA</span> <span class="pre">-e</span> <span class="pre">ToyExpressionData.txt</span> <span class="pre">-m</span> <span class="pre">ToyMotifData.txt</span> <span class="pre">-l</span> <span class="pre">10</span> <span class="pre">-o</span> <span class="pre">ToyOutput</span> <span class="pre">(leaves</span> <span class="pre">out</span> <span class="pre">the</span> <span class="pre">10th</span> <span class="pre">column</span> <span class="pre">of</span> <span class="pre">the</span> <span class="pre">expression</span> <span class="pre">data</span> <span class="pre">when</span> <span class="pre">building</span> <span class="pre">the</span> <span class="pre">network)</span></code></p>
<p>“-j” can be used to jack-knife (random sampling without replacement) the samples in the original expression data. The value indicates the number of samples to use building the network, so this value should be less than the total number of samples and greater than two (since one needs at least three data points to calculate a correlation).</p>
<p>example usage: <code class="docutils literal notranslate"><span class="pre">./PANDA</span> <span class="pre">-e</span> <span class="pre">ToyExpressionData.txt</span> <span class="pre">-m</span> <span class="pre">ToyMotifData.txt</span> <span class="pre">-j</span> <span class="pre">40</span> <span class="pre">-o</span> <span class="pre">ToyOutput</span> <span class="pre">(chooses</span> <span class="pre">40</span> <span class="pre">random</span> <span class="pre">expression</span> <span class="pre">samples/conditions</span> <span class="pre">to</span> <span class="pre">build</span> <span class="pre">network)</span></code></p>
<p>“-r” allows the user to specify several different randomization options.  “-r 1” swaps gene labels, such that each row in the expression data is assigned to a random gene.  “-r 2” swaps condition labels, such that each column in the expression data is assigned to a random weight value in the covariate weight file.  “-r 3” assigns each condition a random weight value and prints out these values.  If used in conjuction with the “-w” option, only conditions that have an original weight value greater than zero are assigned random weight values.</p>
<p>examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">./</span><span class="n">PANDA</span> <span class="o">-</span><span class="n">e</span> <span class="n">ToyExpressionData</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">m</span> <span class="n">ToyMotifData</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">r</span> <span class="mi">1</span> <span class="o">-</span><span class="n">o</span> <span class="n">ToyOutput</span>

<span class="o">./</span><span class="n">PANDA</span> <span class="o">-</span><span class="n">e</span> <span class="n">ToyExpressionData</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">m</span> <span class="n">ToyMotifData</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">w</span> <span class="n">ToyWeights</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">r</span> <span class="mi">2</span> <span class="o">-</span><span class="n">o</span> <span class="n">ToyOutput</span> 
</pre></div>
</div>
<p>“-s” allows the user to seed the random number generator (which otherwise defaults to the current system time). Must be an integer greater than zero.</p>
<p>example: <code class="docutils literal notranslate"><span class="pre">./PANDA</span> <span class="pre">-e</span> <span class="pre">ToyExpressionData.txt</span> <span class="pre">-m</span> <span class="pre">ToyMotifData.txt</span> <span class="pre">-r</span> <span class="pre">1</span> <span class="pre">-s</span> <span class="pre">1000</span> <span class="pre">-o</span> <span class="pre">ToyOutput</span></code></p>
<p>“-v” triggers several verbose output options.  “-v 1” tells PANDA to print out the mean and standard-deviation of the availability and responsibility at each iteration.  “-v 2” tells PANDA to write files containing all three networks (gene-gene co-regulatory, TF-gene regulatory, and TF-TF co-regulating) at both the initial and final states (producing six files).  The gene-gene co-regulatory networks can be very large, so use this option with caution.</p>
<p>example usage: <code class="docutils literal notranslate"><span class="pre">./PANDA</span> <span class="pre">-e</span> <span class="pre">ToyExpressionData.txt</span> <span class="pre">-m</span> <span class="pre">ToyMotifData.txt</span> <span class="pre">-v</span> <span class="pre">1</span> <span class="pre">-o</span> <span class="pre">ToyOutput</span></code></p>
</div>
</div>
<div class="section" id="puma">
<h3>PUMA<a class="headerlink" href="#puma" title="Permalink to this headline">¶</a></h3>
<p>PUMA, or <strong>P</strong>ANDA <strong>U</strong>sing <strong>M</strong>icroRNA <strong>A</strong>ssociations, is a regulatory network reconstruction algorithm that uses message passing to model gene regulatory interactions. PUMA can reconstruct gene regulatory networks using both transcription factors and microRNAs as regulators of mRNA expression levels. This Github repository contains both a MATLAB and a C++ version of PUMA.</p>
<p>The code can be compiled using:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">g</span><span class="o">++</span> <span class="n">PUMA</span><span class="o">.</span><span class="n">c</span> <span class="o">-</span><span class="n">O3</span> <span class="o">-</span><span class="n">o</span> <span class="n">PUMA</span>
</pre></div>
</div>
<p>To run PUMA with the assumption that all regulators can form complexes (estimate <em>responsibility</em> for all regulators, <em>eg</em> a gene regulatory prior with transcription factors only):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">./</span><span class="n">PUMA</span> <span class="o">-</span><span class="n">e</span> <span class="n">ToyExpressionData</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">m</span> <span class="n">ToyMotifData</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">o</span> <span class="n">ToyOutput</span>
</pre></div>
</div>
<p>To tell PUMA to discriminate between regulators that can, and regulators that cannot cannot form complexes (for example a list of microRNAs in <code class="docutils literal notranslate"><span class="pre">miRlist.txt</span></code>), run:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">./</span><span class="n">PUMA</span> <span class="o">-</span><span class="n">e</span> <span class="n">ToyExpressionData</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">m</span> <span class="n">ToyMotifData</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">u</span> <span class="n">miRlist</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">o</span> <span class="n">ToyOutput</span>
</pre></div>
</div>
<p>The PUMA C++ code was based on the PANDA C++ version 2 sourceforge project: https://sourceforge.net/projects/panda-net/.</p>
</div>
</div>
<div class="section" id="troubleshooting">
<h2>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>To report any installation issue or function bug, please report through opening an <a class="reference external" href="https://github.com/netZoo/netZooC/issues">issue</a> on github.</li>
</ul>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../functions/index.html" class="btn btn-neutral float-right" title="Functions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../index.html" class="btn btn-neutral float-left" title="Home" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, netZoo

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>